flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala

```diff
@@ -61,25 +61,108 @@ object AggregateUtil {
+        inputType: RelDataType,
+        outputType: RelDataType,
+        groupings: Array[Int],
+        aggregates:Array[Aggregate[_ <: Any]],
+        aggFieldIndexes:Array[Int])
+    : (MapFunction[Any, Row], RichGroupReduceFunction[Row, Row])= {
+
+      val mapFunction = createAggregateMapFunction(aggregates,
+                        aggFieldIndexes, groupings, inputType)
+
+      // the mapping relation between field index of intermediate aggregate Row and output Row.
+      val groupingOffsetMapping = getGroupKeysMapping(inputType, outputType, groupings)
+
+      // the mapping relation between aggregate function index in list and its corresponding
+      // field index in output Row.
+      val aggOffsetMapping = getAggregateMapping(namedAggregates, outputType)
+
+      if (groupingOffsetMapping.length != groupings.length ||
+        aggOffsetMapping.length != namedAggregates.length) {
+        throw new TableException("Could not find output field in input data type " +
+          "or aggregate functions.")
+      }
+
+      val allPartialAggregate = aggregates.map(_.supportPartial).forall(x => x)
+
+      val intermediateRowArity = groupings.length +
+                        aggregates.map(_.intermediateDataType.length).sum
+
+      val reduceGroupFunction =
+        if (allPartialAggregate) {
+          new AggregateReduceCombineFunction(
+            aggregates,
+            groupingOffsetMapping,
+            aggOffsetMapping,
+            intermediateRowArity,
+            outputType.getFieldCount)
+        }
+        else {
+          new AggregateReduceGroupFunction(
+            aggregates,
+            groupingOffsetMapping,
+            aggOffsetMapping,
+            intermediateRowArity,
+            outputType.getFieldCount)
+        }
+
+      (mapFunction, reduceGroupFunction)
+  }
+
+  /**
+    * Create Flink operator functions for Incremental aggregates.
+    * It includes 2 implementations of Flink operator functions:
+    * [[org.apache.flink.api.common.functions.MapFunction]] and
+    * [[org.apache.flink.api.common.functions.ReduceFunction]]
+    * The output of [[org.apache.flink.api.common.functions.MapFunction]] contains the
+    * intermediate aggregate values of all aggregate function, it's stored in Row by the following
+    * format:
+    *
+    * {{{
+    *                   avg(x) aggOffsetInRow = 2          count(z) aggOffsetInRow = 5
+    *                             |                          |
+    *                             v                          v
+    *        +---------+---------+--------+--------+--------+--------+
+    *        |groupKey1|groupKey2|  sum1  | count1 |  sum2  | count2 |
+    *        +---------+---------+--------+--------+--------+--------+
+    *                                              ^
+    *                                              |
+    *                               sum(y) aggOffsetInRow = 4
+    * }}}
+    *
+    */
```

Comment:
It would be better to describe the meaning of the return value. Especially  `Array[(Int, Int)],Array[(Int, Int)],Int`.